### PR TITLE
Add feature: forking and saving

### LINKED ISSUE #27: Add ability to fork a saved config
Add a property "forks", and "forkedFrom" to configs in the db.

Make a new API route.

```
POST /fork/:id
```

When someone hits this route.

Get the config with `id`, let's call this `source`.

Create a new config with:

```
forkedFrom: source.id
config: source.config
```

Then update `source` and append the id of the new row to `forks`.

### LINKED ISSUE #28: Render fork information in UI
See #27

Should include:

- link to parent
- the # of forks
- possibility to list forks

### PR DESCRIPTION
**Pull request by garrettluu**
Closes #28, addresses #27 

This PR adds UI elements to fork and save a config. I ran into several issues as I was doing this, stemming from the problem of not being able to get the `id` of the current config (since that isn't stored in the state anywhere). So I propose a new process for forking:
- Forking is not available until user has saved the current config
  - If config is not saved, we do not have a blob to fork from
- Once the user saves the config, the "fork" button appears
  - Create a new blob or update the current blob and store the `id` somewhere in the application state
- Once the user forks, we create a `newBlob` identical to the current `blob` but with a different `id`
  - Add `newBlob.id` to `blob.forks[]` and make `newBlob.forkedFrom` point to `blob.id` (this is done in backend)

~**TODO**~ done!
- [x] Update `REPLState` to allow for saving and forking
- [x] Store `id` in application state, or change the url after saving and get `id` using `extractID`
- [x] Coordinate with backend for API endpoints and data
- [x] Allow for viewing of forks
- [x] Test forking with real data

### CODE REVIEW

File: src/state/detectShare.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ function extractID() {
 
   // Attempt to capture :key from http://example.com/share/:key/
   const pathParts = window.location.pathname.split('/');
-  if (pathParts[1] == 'share' && pathParts[2]) {
+  if (pathParts[1] === 'share' && pathParts[2]) {
```
Comment (ianjennings) [inline src/state/detectShare.js]: Good practice, but unlikely the type will change here.

File: src/state/REPLState.js [hunk 2]
```diff
@@ -75,21 +76,46 @@ class REPLState {
...
+        message = await this.New();
+        setId(message.id);
+      } else {
+        message = await this.Save(id);
+      }
+
+      // https://stackoverflow.com/questions/6941533/get-protocol-domain-and-port-from-url
+      return (
+        window.location.href.split("/").slice(0, 3).join("/") + message.url
+      );
```
Comment (ianjennings) [inline src/state/REPLState.js]: I don't believe `()` is needed.

File: src/state/REPLState.js [hunk 3]
```diff
@@ -75,21 +76,46 @@ class REPLState {
...
         return decodeBase64(configs);
-      })
+      }),
     );
   }
 
   /**
    * Link gets the sharing the sharing link
-   * for the given REPL state.
+   * for the given REPL state and updates the id
```
Comment (ianjennings) [inline src/state/REPLState.js:90]: <3

File: src/state/REPLState.js [hunk 4]
```diff
@@ -75,21 +76,46 @@ class REPLState {
    */
   static Decode(encodedState) {
     let jsonState = JSON.parse(encodedState);
+    console.log(jsonState);
```
Comment (ianjennings) [inline src/state/REPLState.js]: remove before merge

**Comment by garrettluu** [comment]
Finally added displaying forks! Not sure how much I should style this since the UI and layout are probably gonna change soon.
image

**Comment by garrettluu** [comment]
Tested forking with the backend and it seems to work and create new blobs, but I would also like to track the forks of the parent in an array in the blob (this is a backend task @mohammedsahl). 

If we start with this blob
```json
{
  "id": "example1",
  "forkedFrom": "",
  "forks": []
}
```
and we fork it to generate this blob
```json
{
  "id": "example2",
  "forkedFrom": "example1",
  "forks": []
}
```
the `forks` array of `example1` should contain `example2` as an element:
`"forks": ["example2"]`

Currently, the `forks` array seems to always be empty or null.

**Comment by mohammedsahl** [comment]
> Currently, the `forks` array seems to always be empty or null.

Okay so this seems to be an issue(?) with async DB updates. If you `GET` `example1` after forking it should have `forks : ["example2"]`

Edit: It appears that `forks: [...]` is not part of the `res` body when `GET`ing a single blob, I'll add that right away

**Comment by garrettluu** [comment]
Tested with real data from backend and everything seems to work! 😀 
image

File: src/components/App.js [hunk 5]
```diff
@@ -3,11 +3,13 @@ import * as Babel from "@babel/standalone";
 
 import { CustomPlugin } from "./CustomPlugin";
 import { MainMenu } from "./MainMenu";
+import { Forks } from "./Forks";
 import { Input } from "./Input";
 import { Output } from "./Output";
 import { gzipSize } from "../gzip";
 import { Root } from "./styles";
 import { useDebounce } from "../utils/useDebounce";
+import REPLState from "../state/REPLState";
```
Comment (wlawt) [inline src/components/App.js]: Can we remove the imports that aren't being used anymore (along with the comments on line 95-98)?

File: src/components/Forks.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import React, { useState } from 'react';
+import REPLState from '../state/REPLState';
```
Comment (wlawt) [inline src/components/Forks.js]: `useState` and `REPLState` aren't being used either

File: src/components/MainMenu.js [hunk 7]
```diff
@@ -11,9 +53,15 @@ export function MainMenu({
...
   enableCustomPlugin,
+  id,
+  setId,
+  setForksVisible,
+  forks,
+  setForks,
 }) {
   const [shareLink, setShareLink] = React.useState("");
   const [showShareLink, setShowShareLink] = React.useState(false);
+  console.log(id);
```
Comment (wlawt) [inline src/components/MainMenu.js]: Before merge, we should remove some of the `console.log`

File: src/components/MainMenu.js [hunk 8]
```diff
@@ -1,8 +1,50 @@
...
+//   const url = `/api/v1/blobs/get-blob/${ID}`;
+//   try {
+//     const resp = await fetch(url);
+//     return resp.json()["forks"];
+//   } catch (err) {
+//     console.error(err);
+//     return null;
+//   }
+// }
+
```
Comment (wlawt) [inline src/components/MainMenu.js]: Do we still need this for future references?

File: src/components/App.js [hunk 9]
```diff
@@ -75,17 +76,26 @@ function registerDefaultPlugins() {
   );
 }
```
Comment (wlawt) [inline src/components/App.js:79]: I think `import VizOutput from "./AST/Viz";` was accidentally deleted

File: src/components/App.js [hunk 10]
```diff
@@ -75,17 +76,26 @@ function registerDefaultPlugins() {
...
   const [gzip, setGzip] = useState(null);
   const debouncedSource = useDebounce(source, 125);
 
+  const [forksVisible, setForksVisible] = useState(false);
+  const [forks, setForks] = useState([]);
+
+  // (async () => {
+  //   const blob = await REPLState.GetBlob(id) ?? null;
+  //   setBlob(blob);
+  // })();
```
Comment (wlawt) [inline src/components/App.js]: We can also remove this, if not needed

**Comment by wlawt** [comment]
We should also update the docs about how the forking logic works just to clarify.

The current process is: 
- Add custom plugin -> save -> share -> copy link -> fork -> paste link that was copied -> click the number

**Comment by ianjennings** [comment]
```
./src/components/MainMenu.js
Module not found: Can't resolve './ForkModal' in '/home/ianjennings/Development-Unix/babel-sandbox/src/components'
```

I don't see this file in my local copy, do you need to commit it?

**Pull request merged**